File: kotlin/src/hu/klenium/tetris/util/PeriodicTask.kt
package hu.klenium.tetris.util

import java.util.*
import kotlin.concurrent.timer

class PeriodicTask(private val task: () -> Unit, delay: Int) {
    private val time = delay.toLong()
    private var clock: Timer? = null
    fun start() {
        clock = clock ?: timer(
            daemon = true,
            initialDelay = time,
            period = time,
            action = { task.invoke() }
        )
    }
    fun stop() {
        clock?.cancel()
        clock = null
    }
    fun reset() {
        stop()
        start()
    }
}

File: csharp/TetrisGameBase/util/PeriodicTask.cs
﻿using System;
using System.Timers;

namespace hu.klenium.tetris.util
{
    public class PeriodicTask
    {
        private Timer timer = null;
        public PeriodicTask(Action task, int delay)
        {
            timer = new Timer();
            timer.Interval = delay;
            timer.AutoReset = true;
            timer.Elapsed += (t, e) => task.Invoke();
        }
        public void Start()
        {
            timer.Enabled = true;
        }
        public void Stop()
        {
            timer.Enabled = false;
        }
        public void Reset()
        {
            Stop();
            Start();
        }
    }
}